python_modules/dagster/dagster_tests/definitions_tests/test_sensor_invocation.py

```diff
@@ -173,22 +176,110 @@ def test_instance_access_with_mock():
+
+
+def test_validated_partitions():
+    @job(partitions_def=StaticPartitionsDefinition(["foo", "bar"]))
+    def my_job():
+        pass
+
+    @sensor(job=my_job)
+    def invalid_req_sensor():
+        return RunRequest(partition_key="nonexistent")
+
+    @sensor(job=my_job)
+    def valid_req_sensor():
+        return RunRequest(partition_key="foo", tags={"yay": "yay!"})
+
+    @repository
+    def my_repo():
+        return [my_job, invalid_req_sensor, valid_req_sensor]
+
+    with pytest.raises(DagsterInvariantViolationError, match="Must provide repository def"):
+        with build_sensor_context() as context:
+            invalid_req_sensor.evaluate_tick(context)
+
+    with pytest.raises(DagsterUnknownPartitionError, match="Could not find a partition"):
+        with build_sensor_context(repository_def=my_repo) as context:
+            invalid_req_sensor.evaluate_tick(context)
+
+    with build_sensor_context(repository_def=my_repo) as context:
+        run_requests = valid_req_sensor.evaluate_tick(context).run_requests
+        assert len(run_requests) == 1
+        run_request = run_requests[0]
+        assert run_request.partition_key == "foo"
+        assert run_request.run_config == {}
+        assert run_request.tags.get(PARTITION_NAME_TAG) == "foo"
+        assert run_request.tags.get("yay") == "yay!"
+
+
+def test_partitioned_config_run_request():
+    def partition_fn(partition_key: str):
+        return {"ops": {"my_op": {"config": {"partition": partition_key}}}}
+
+    @static_partitioned_config(partition_keys=["a", "b", "c", "d"])
+    def my_partitioned_config(partition_key: str):
+        return partition_fn(partition_key)
+
+    @op
+    def my_op():
+        pass
+
+    @job(config=my_partitioned_config)
+    def my_job():
+        my_op()
+
+    @sensor(job=my_job)
+    def valid_req_sensor():
+        return RunRequest(partition_key="a", tags={"yay": "yay!"})
+
+    @sensor(job=my_job)
+    def invalid_req_sensor():
+        return RunRequest(partition_key="nonexistent")
+
+    @sensor(job_name="my_job")
+    def job_str_target_sensor():
+        return RunRequest(partition_key="a", tags={"yay": "yay!"})
+
+    @sensor(job_name="my_job")
+    def invalid_job_str_target_sensor():
+        return RunRequest(partition_key="invalid")
+
+    @repository
+    def my_repo():
+        return [
+            my_job,
+            valid_req_sensor,
+            invalid_req_sensor,
+            job_str_target_sensor,
+            invalid_job_str_target_sensor,
+        ]
+
+    with build_sensor_context(repository_def=my_repo) as context:
+        for valid_sensor in [valid_req_sensor, job_str_target_sensor]:
+            run_requests = valid_sensor.evaluate_tick(context).run_requests
+            assert len(run_requests) == 1
+            run_request = run_requests[0]
+            assert run_request.run_config == partition_fn("a")
+            assert run_request.tags.get(PARTITION_NAME_TAG) == "a"
```

Comment:
Ah, so if you do direct invocation
```
valid_sensor(context)
```
you'll just get back the run request returned by the user.

`evaluate_tick` is where we do things like resolving the asset selection, checking for valid target job, etc. so that's where a partitioned run request is resolved and the tags are added.

Reply:
do we modify the RunRequest in evaluate_tick in other ways? do we add the dagster/sensor_name tag?

Reply:
No, we add the sensor name tag in the daemon

Reply:
hm... actually it all seems pretty fragmented at the moment. for schedules we add the schedule tag in `evaluate_tick`, but in sensors we do it in the daemon. 

I'm happy to move all of this logic into the daemon to unify this

Reply:
On second thought, resolving the run request requires calling out to user code to fetch the run config per partition. If we did this in the daemon it'd require a gRPC call, I think I'd prefer resolving the run request in `evaluate_tick` instead

Reply:
Got it - makes sense